### PR TITLE
Add instruction for custom istio gateway

### PR DESCRIPTION
**Pull request by nak3**
This patch adds the instruction to use custom Istio Gateway.

Current doc has a similar instruction "customize Istio Gateway
Service". This doc is very similar but for "customize Istio Gateway".

Fix https://github.com/knative/docs/issues/3524

### CODE REVIEW

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 1]
```diff
@@ -92,3 +79,63 @@ For the service above, it should be updated to:
...
+    - "*"
+EOF
+```
+
+!!! note
+    Replace the label selector `istio: ingressgateway` with the label of your service.
+
+### Step 2: Update Gateway Configmap
+
+Update gateway configmap `config-istio` under `knative-serving`
```
Comment (snneji) [inline docs/serving/setting-up-custom-ingress-gateway.md:119]:
```suggestion
Update the gateway ConfigMap `config-istio` under the `knative-serving`
```

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 2]
```diff
@@ -92,3 +79,63 @@ For the service above, it should be updated to:
...
+namespace:
```
Comment (snneji) [inline docs/serving/setting-up-custom-ingress-gateway.md:120]:
```suggestion
namespace by running the command:
```

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 3]
```diff
@@ -92,3 +79,63 @@ For the service above, it should be updated to:
...
+```
+
+For the gateway above, it should be updated to:
+
+```
+gateway.custom-ns.knative-custom-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
+```
+
+The configuration format should be `gateway.GATEWAY_NAMESPACE.GATEWAY_NAME`.
+The `GATEWAY_NAMESPACE` is optional. when it is omitted, the system will search for
```
Comment (snneji) [inline docs/serving/setting-up-custom-ingress-gateway.md:140]:
```suggestion
The configuration format should be `gateway.<gateway-namespace>.<gateway-name>`.
`<gateway-namespace>` is optional. When it is omitted, the system searches for
```

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 4]
```diff
@@ -92,3 +79,63 @@ For the service above, it should be updated to:
 ```
 gateway.knative-serving.knative-ingress-gateway: custom-ingressgateway.custom-ns.svc.cluster.local
 ```
+
+## Replace the knative-ingress-gateway gateway
```
Comment (snneji) [inline docs/serving/setting-up-custom-ingress-gateway.md:83]:
```suggestion
## Replace the `knative-ingress-gateway` gateway
```

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 5]
```diff
@@ -11,9 +11,11 @@ Knative uses a shared ingress Gateway to serve all incoming traffic within
 Knative service mesh, which is the `knative-ingress-gateway` Gateway under
 the `knative-serving` namespace. By default, we use Istio gateway service
 `istio-ingressgateway` under `istio-system` namespace as its underlying service.
-You can replace the service with that of your own as follows.
+You can replace the service and the gateway with that of your own as follows.
 
-## Step 1: Create Gateway Service and Deployment Instance
+## Replace the default `istio-ingressgateway` service
+
+#### Step 1: Create Gateway Service and Deployment Instance
```
[the remark this answers is not in the document]
Comment (snneji) [reply docs/serving/setting-up-custom-ingress-gateway.md:18]: The style in the Knative docs is to use sentence case in headings.

**Review by snneji — CHANGES_REQUESTED** [review]
Hi @nak3, really good information in this PR. I've suggested a few changes so that the new information is formatted using Knative style rules, and is easier to understand.

File: docs/serving/setting-up-custom-ingress-gateway.md (hunk 3 above)
Comment (snneji) [reply docs/serving/setting-up-custom-ingress-gateway.md:140]: Knative style rules say that placeholders should be in the format `<placeholder-name>` and not `PLACEHOLDER_NAME`.

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 6]
```diff
@@ -92,3 +79,63 @@ For the service above, it should be updated to:
...
+
+For the gateway above, it should be updated to:
+
+```
+gateway.custom-ns.knative-custom-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
+```
+
+The configuration format should be `gateway.GATEWAY_NAMESPACE.GATEWAY_NAME`.
+The `GATEWAY_NAMESPACE` is optional. when it is omitted, the system will search for
+the gateway in the serving system namespace `knative-serving`.
```
[the remark this answers is not in the document]
Comment (nak3) [reply docs/serving/setting-up-custom-ingress-gateway.md]: Sure! Let me try to change it.

**Comment by nak3** [comment]
Thank you @snneji ! Updated.

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 7]
```diff
@@ -71,24 +58,112 @@ istio: custom-gateway
...
+         ################################
+         #                              #
+         #    EXAMPLE CONFIGURATION     #
+         #                              #
+         ################################
+         # ...
+         gateway.knative-serving.knative-ingress-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
+     ```
+
+1. Edit the file to replace the `istio-ingressgateway.istio-system.svc.cluster.local` field with
```
Comment (ZhiminXiang) [inline docs/serving/setting-up-custom-ingress-gateway.md:85]: I feel "replace" might be confusing as we actually "add" a new field into "data" section.

File: docs/serving/setting-up-custom-ingress-gateway.md [hunk 8]
```diff
@@ -71,24 +58,112 @@ istio: custom-gateway
...
+         ################################
+         #                              #
+         #    EXAMPLE CONFIGURATION     #
+         #                              #
+         ################################
+         # ...
+         gateway.knative-serving.knative-ingress-gateway: "istio-ingressgateway.istio-system.svc.cluster.local"
+     ```
+
+1. Edit the file to replace the `gateway.knative-serving.knative-ingress-gateway` field with
```
Comment (ZhiminXiang) [inline docs/serving/setting-up-custom-ingress-gateway.md:155]: same here. we may not want to use "replace".

**Review by ZhiminXiang — COMMENTED** [review]
thanks for updating this doc @nak3!

**Comment by nak3** [comment]
Thank you Zhimin! updated.

**Comment by nak3** [comment]
@snneji Friendly ping. This is ready for merge, I think :smile_cat:

**Comment by omerbensaadon** [comment]
sorry for the delay @nak3, hit us in the #docs channel in the future 👍🏼

**Pull request merged**